src/linux/device_info.cpp

```diff
@@ -0,0 +1,127 @@
+#include "device_info.h"
+#include <QFile>
+#include <QTextStream>
+#include <QJsonObject>
+#include <QDebug>
+
+device_info::device_info()
+{
+    _is_raspberry_pi = false;
+    _revision = "";
+    _device_code = "";
+    _tags = {};
+    _tags_set = false;
+    _hardware_name = "";
+    
+}
+
+bool device_info::determineHardware()
+{
+    // Attempt to open /proc/cpuinfo
+    QFile f("/proc/cpuinfo");
+    if (f.open(QFile::ReadOnly | QFile::Text)) {
+        QTextStream s(&f);
+        while (!s.atEnd()) {
+            QString line = s.readLine();
+            if (line.startsWith("Revision")) {
+                _revision = line.section(':', 1).trimmed();
+                bool ok;
+                uint32_t value_int = _revision.toUInt(&ok, 16);
+
+                // Device type is 8 bits, stored in bits 4-11
+                const uint32_t MASK = 0xFF << 4;
+                uint32_t isolated_bits = value_int & MASK;
+                uint32_t extracted_value = isolated_bits >> 4;
+
+                // convert final value back to HEX
+                std::stringstream ss;
+                ss << std::hex << extracted_value;
+                QString extracted_hex_string = QString::fromStdString(ss.str());
+                _device_code = extracted_hex_string;
+
+                // Make a Human Readable Equivalent of this device_code
+                std::unordered_map<uint32_t, std::string> pi_revision_map = {
+                    {0x00, "Raspberry Pi 1"},
+                    {0x01, "Raspberry Pi 1"},
+                    {0x02, "Raspberry Pi 1"},
+                    {0x03, "Raspberry Pi 1"},
+                    {0x04, "Raspberry Pi 2"},
+                    {0x06, "Raspberry Pi 1"},
+                    {0x08, "Raspberry Pi 3"},
+                    {0x09, "Raspberry Pi Zero"},
+                    {0x0a, "Raspberry Pi 3"},
+                    {0x0c, "Raspberry Pi Zero"},
+                    {0x0d, "Raspberry Pi 3"},
+                    {0x0e, "Raspberry Pi 3"},
+                    {0x10, "Raspberry Pi 3"},
+                    {0x11, "Raspberry Pi 4"},
+                    {0x12, "Raspberry Pi Zero 2 W"},
+                    {0x13, "Raspberry Pi 4"},
+                    {0x14, "Raspberry Pi 4"},
+                    {0x15, "Raspberry Pi 4"},

```

Comment:
Just for reference and readability purposes, it might be nice to have comments here clarifying the _actual_ Pi model? E.g.
```suggestion
                    {0x13, "Raspberry Pi 4"}, // 400
                    {0x14, "Raspberry Pi 4"}, // CM4
                    {0x15, "Raspberry Pi 4"}, // CM4S
```